clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java

```diff
@@ -87,15 +102,22 @@ public MetadataResponse(List<Node> brokers, List<TopicMetadata> topicMetadata) {
             broker.set(NODE_ID_KEY_NAME, node.id());
             broker.set(HOST_KEY_NAME, node.host());
             broker.set(PORT_KEY_NAME, node.port());
+            if (broker.hasField(RACK_KEY_NAME))
+                broker.set(RACK_KEY_NAME, node.rack());
             brokerArray.add(broker);
         }
         struct.set(BROKERS_KEY_NAME, brokerArray.toArray());
 
+        if (struct.hasField(CONTROLLER_ID_KEY_NAME))
+            struct.set(CONTROLLER_ID_KEY_NAME, controllerId);
```

Comment:
I think it would be useful to add a comment here and when we set `RACK_KEY_NAME` saying stating the version they were added in (we do that in other request/response classes).